### PR TITLE
DLEQ implementation

### PR DESCRIPTION
**Pull request by Lederstrumpf**

### CODE REVIEW

**Comment by h4sh3d** [comment]
Do you need the nightly feature from secpfun? It is enabled by default.

**Comment by h4sh3d** [comment]
Looking at the doc nightly is just for optimized arithmetic, IIRC it does not change the API, if so I’m for deactivating it instead of removing stable channel

**Comment by Lederstrumpf** [comment]
> Looking at the doc nightly is just for optimized arithmetic, IIRC it does not change the API, if so I’m for deactivating it instead of removing stable channel

good catch! I've consequently removed it from the features - so now temporary hack of removing stable channel is gone.

File: src/crypto/dleq.rs [hunk 1]
```diff
@@ -0,0 +1,1047 @@
...
+use bitvec::{order::Lsb0, prelude::BitSlice};
+use curve25519_dalek::{
+    constants::ED25519_BASEPOINT_POINT as G, edwards::CompressedEdwardsY as ed25519PointCompressed,
+    edwards::EdwardsPoint as ed25519Point, scalar::Scalar as ed25519Scalar,
+};
+
+const ENTROPY: bool = true;
+
+fn _max_ed25519() -> u256 {
+    (u256::from(2u32) << 252) + 27742317777372353535851937790883648493u128
```
Comment (h4sh3d) [inline src/crypto/dleq.rs:20]: There is a typo here: `2 = 0b0000_0010`, left shifting it by 252 ends up to `2^253 + ...`

Did some quick test:
```sage
sage: s=2**252+27742317777372353535851937790883648493
sage: s
7237005577332262213973186563042994240857116359379907606001950938285454250989
sage: s.hex()
'1000000000000000000000000000000014def9dea2f79cd65812631a5cf5d3ed'
```
and the rust code
```rust
let num = (u256::from(2u32) << 252) + 27742317777372353535851937790883648493u128;
println!("{}", num);
0x2000000000000000000000000000000014def9dea2f79cd65812631a5cf5d3ed
let num = (u256::from(1u32) << 252) + 27742317777372353535851937790883648493u128;
println!("{}", num);
0x1000000000000000000000000000000014def9dea2f79cd65812631a5cf5d3ed
```

File: src/crypto/dleq.rs [hunk 2]
```diff
@@ -0,0 +1,1047 @@
...
+    ed25519Scalar::from(8u8) * hash_to_curve
+}
+
+#[cfg(feature = "experimental")]
+use ecdsa_fun::fun::{Point as secp256k1Point, Scalar as secp256k1Scalar, G as H};
+#[cfg(feature = "experimental")]
+use secp256kfun::{g, marker::*, s as sc};
+use sha2::Digest;
+
+fn _max_secp256k1() -> u256 {
```
Comment (h4sh3d) [inline src/crypto/dleq.rs:47]: LGTM, tested with SageMath, bytes and arithmetic validated.

File: src/crypto/dleq.rs [hunk 3]
```diff
@@ -0,0 +1,1047 @@
...
+            .unwrap();
+        if !ecdsa_fun::ECDSA::verify(&ecdsa, &self.xH_p, &pok_1_message_hash, &self.pok_1) {
+            return Err(crypto::Error::InvalidProofOfKnowledge);
+        }
+
+        // Everything ok
+        Ok(())
+    }
+}
+
```
Comment (h4sh3d) [inline src/crypto/dleq.rs:959]: Maybe add the convention of test modules:
```rust
#[cfg(test)]
mod tests {
    ...
}
```

**Review by h4sh3d — COMMENTED** [review]
Didn't went deep in all the details, but LGTM overall (with comments already exchanged on IRC between @kayabaNerve and @Lederstrumpf)

File: src/crypto/dleq.rs (hunk 1 above)
Comment (Lederstrumpf) [reply src/crypto/dleq.rs]: thanks! fixed.

File: src/crypto/dleq.rs (hunk 3 above)
Comment (Lederstrumpf) [reply src/crypto/dleq.rs:960]: I'll do this in #144 - otherwise too much rebasing.

File: src/crypto/dleq.rs [hunk 4]
```diff
@@ -0,0 +1,1048 @@
...
+fn reverse_endianness(bytes: &[u8; 32]) -> [u8; 32] {
+    let mut bytes_rev = *bytes;
+    bytes_rev.reverse();
+    bytes_rev
+}
+
+// matches https://github.com/monero-project/monero/blob/9414194b1e47730843e4dbbd4214bf72d3540cf9/src/ringct/rctTypes.h#L454
+// i.e. hash-to-curve of G as in https://github.com/monero-project/mininero/blob/master/mininero.py#L305-L323
+// TODO: this is disgusting and must be removed asap
```
Comment (TheCharlatan) [inline src/crypto/dleq.rs:31]: Can this TODO be removed now?

**Review by TheCharlatan — APPROVED** [review]

File: src/crypto/dleq.rs (hunk 4 above)
Comment (Lederstrumpf) [reply src/crypto/dleq.rs:31]: No - I'll address that in the followup PR. Disgusting := inefficient.

**Pull request merged**